python/taichi/lang/ast/ast_transformer_utils.py

```diff
@@ -177,7 +178,10 @@ def get_var_by_name(self, name):
         for s in reversed(self.local_scopes):
             if name in s:
                 return s[name]
-        return self.global_vars.get(name)
+        var = self.global_vars.get(name)
+        if var is not None:
+            return var
+        return getattr(builtins, name, None)
```

Comment:
Is there any case the value is `None`?

Maybe changed to:
```python
if name in self.global_vars:
    return self.global_vars[name]
return getattr(builtins, name, None)
```

Further, is returning `None` correct when name is not found?

Reply:
> Is there any case the value is `None`?
> 
> Maybe changed to:
> 
> ```python
> if name in self.global_vars:
>     return self.global_vars[name]
> return getattr(builtins, name, None)
> ```
> 
> Further, is returning `None` correct when name is not found?

maybe?
```python
return self.global_vars.get(name, getattr(builtins, name, None))
```

Reply:
@frostming It should raise error when nothing is found, and it will be done in #3951 which depends on this PR (this PR fixes the bug which causes the test of #3951 to fail).